File: backend/app/models/job.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum
from datetime import datetime
import uuid


class JobStatus(str, Enum):
    open = "open"
    closed = "closed"
    filled = "filled"


class BudgetRange(BaseModel):
    min: int = Field(ge=0)
    max: int = Field(ge=0)

    def model_post_init(self, __context):
        if self.min > self.max:
            raise ValueError("min は max 以下である必要があります")


class RequiredSpecs(BaseModel):
    skills: Optional[list[str]] = None
    min_uptime: Optional[float] = Field(None, ge=0, le=100)
    max_response_ms: Optional[float] = Field(None, ge=0)
    pricing_model: Optional[str] = None
    other: Optional[str] = Field(None, max_length=1000)


class JobBase(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    problem_statement: str = Field(min_length=1, max_length=5000)
    budget_range: Optional[BudgetRange] = None
    required_specs: Optional[RequiredSpecs] = None


class JobCreate(JobBase):
    pass


class JobRawInput(BaseModel):
    raw_description: str = Field(min_length=10, max_length=10000)


class Job(JobBase):
    id: uuid.UUID
    company_id: uuid.UUID
    status: JobStatus
    created_at: datetime

    model_config = {"from_attributes": True}

File: src/schemas/job.ts
import { z } from "zod";

export const JobStatusSchema = z.enum(["open", "closed", "filled"]);
export type JobStatus = z.infer<typeof JobStatusSchema>;

export const BudgetRangeSchema = z.object({
  min: z.number().int().nonnegative(),
  max: z.number().int().nonnegative(),
}).refine((data) => data.min <= data.max, {
  message: "最小予算は最大予算以下である必要があります",
});
export type BudgetRange = z.infer<typeof BudgetRangeSchema>;

export const RequiredSpecsSchema = z.object({
  skills: z.array(z.string()).optional(),
  min_uptime: z.number().min(0).max(100).optional(),
  max_response_ms: z.number().nonnegative().optional(),
  pricing_model: z.enum(["subscription", "usage_based", "any"]).optional(),
  other: z.string().max(1000).optional(),
});
export type RequiredSpecs = z.infer<typeof RequiredSpecsSchema>;

export const JobSchema = z.object({
  id: z.string().uuid(),
  company_id: z.string().uuid(),
  title: z.string().min(1).max(200),
  problem_statement: z.string().min(1).max(5000),
  budget_range: BudgetRangeSchema.nullable(),
  required_specs: RequiredSpecsSchema.nullable(),
  status: JobStatusSchema,
  created_at: z.string().datetime(),
});
export type Job = z.infer<typeof JobSchema>;

export const CreateJobSchema = JobSchema.omit({
  id: true,
  company_id: true,
  status: true,
  created_at: true,
});
export type CreateJobInput = z.infer<typeof CreateJobSchema>;

export const UpdateJobSchema = CreateJobSchema.partial().extend({
  status: JobStatusSchema.optional(),
});
export type UpdateJobInput = z.infer<typeof UpdateJobSchema>;

// 企業が雑に書いた課題 → LLMが構造化するためのスキーマ
export const JobRawInputSchema = z.object({
  raw_description: z.string().min(10).max(10000),
});
export type JobRawInput = z.infer<typeof JobRawInputSchema>;
